adding more checkers

allrob23 opened the pull request:
Hey there,

We’re working on a research project in runtime verification and have built a set of checkers that we thought might be useful for DyLin. This PR adds 19 new checkers. The description of each checker is available as a comment at the top of its file.

Let me know what you think. I'm Happy to tweak anything if needed!

AryazE commented:
Thanks a lot for the tests, they look very good. Now I have a better understanding of what each checker is looking for.
Could you also address the comments by copilot?

I am a bit unsure about adding the checkers that raise exceptions during runtime. Since DyLin is a dynamic linter, and we need to execute the code to find the issues, the Python interpreter already warns about these errors at the correct location, which makes these checkers redundant.
My suggestion is to remove them, but let me know what you think.

Thanks again for the nice contribution!

AryazE commented on src/dylin/select_checkers.py:
```diff
@@ -91,6 +91,101 @@
         "analysis": "dylin.analyses.GradientAnalysis.GradientAnalysis",
         "aliases": ["M-28"],
     },
+    "TE-02": {
```
I think we can move most of the checkers to either the "Standard Library" category (SL-*) or the "Programming Constructs" (PC-*). The rest could remain uncategorized until we have something for them.

AryazE commented:
I'm also curious about how you came up with these checkers.

mohas2 commented:
Thanks. glad to hear the checkers were helpful!

We developed them as part of an ongoing research project and will share more details once the work is published.

I’ll address the points you raised above shortly.

allrob23 commented:
Thanks. glad to hear the checkers were helpful!

We developed them as part of an ongoing research project and will share more details once the work is published.

I’ll address the points you raised above shortly.

allrob23 commented:
Hi @AryazE, I've addressed all the copilot comments and removed all the checkers that python already handles. However, the following are still failing for the following reasons:

```
CreateWidgetOnSameFrameCanvas -> There is no display device in the CI.
NLTK_RegexpTokenizerCapturingParentheses -> same as above.
PyDocs_MustOnlyAddSynchronizableDataToSharedList -> We will debug further to understand why it's failing.
Console_CloseWriter ->  because the test closes the stdout which is needed for the rest of the test suite.
Console_CloseErrorWriter -> same as above.
```

We're happy to discuss these more if needed

AryazE commented:
The changes look good to me.
To avoid the issues you mentioned and move forward faster, I suggest to remove the problematic ones from this PR, and we can open another PR for them to work on those. They might need some changes to the testing harness (the CloseWriter checkers specifically).
Thanks

allrob23 commented:
I removed all specs with broken tests, now all tests are passing.
